core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,461 @@
+   *           attribute values for that expression
+   */
+  def executeQueryPlan(workspaceId: UUID,
+                       entityType: String,
+                       entityName: String,
+                       rootEntityType: String,
+                       plan: QueryPlan,
+                       entityLookups: Seq[ExpressionLookup] = Seq.empty
+  )(implicit
+    executionContext: ExecutionContext
+  ): ReadAction[Seq[ExpressionAndResult]] = {
+    val queryAction =
+      if (plan.relationChain.isEmpty) {
+        repository.queries.getEntity(workspaceId, entityType, entityName).map {
+          case Some(entity) => Map(entity.name -> Seq(entity)) // Wrap the entity in a Seq to match the expected type
+          case None =>
+            throw new RawlsExceptionWithErrorReport(
+              ErrorReport(StatusCodes.NotFound, s"Entity of type $entityType with name $entityName not found.")
+            )
+        }
+      } else {
+        repository.queries.queryRelatedRecordsWithRelationChain(workspaceId, entityType, entityName, plan.relationChain)
+      }
+
+    queryAction.map { entityRecords =>
+      // Validate entity types if we have entityLookups
+      if (entityLookups.nonEmpty && entityRecords.nonEmpty) {
+        val actualEntityTypes = entityRecords.values.flatten.map(_.entityType).toSet
+        if (actualEntityTypes.nonEmpty && !actualEntityTypes.contains(rootEntityType)) {
+          val actualTypesStr = actualEntityTypes.mkString(", ")
+          throw new RawlsExceptionWithErrorReport(
+            ErrorReport(
+              StatusCodes.BadRequest,
+              s"The expression in your SubmissionRequest matched only entities of the wrong type. " +
+                s"(Expected type $rootEntityType, but got $actualTypesStr.)"
+            )
+          )
+        }
+      }
+
+      // For each expression in this query plan, create an ExpressionAndResult
+      plan.expressionMappings.toSeq.flatMap { case (expression, attributeNames) =>
+        attributeNames.map { attrName =>
+          val attributeName = AttributeName.fromDelimitedName(attrName)
+
+          val entityToAttributeValues = if (entityType == rootEntityType) {
+            // Group all results under the original entityName since we want results grouped by the starting entity type
+            val allAttrs: Seq[AttributeValue] = entityRecords.values.flatten.toSeq.flatMap { record =>
+              val attributeNameToCheck =
+                AttributeName.withDefaultNS(record.entityType + Attributable.entityIdAttributeSuffix)
+              record.toEntity.attributes.get(attributeName) match {
+                case Some(avl: AttributeValueList) =>
+                  avl.list.toSeq
+                case Some(av: AttributeValue) =>
+                  Seq(av)
+                case _ if attributeName == attributeNameToCheck =>
+                  Seq(AttributeString(record.name))
+                case _ =>
+                  Seq.empty
+              }
```

Comment:
nit: when asking for the entity name, it's pointless to look in the attribute map:
```suggestion
              // If asking for the entity's name, just return that; else retrieve the attribute values
              if (attributeName == AttributeName.withDefaultNS(record.entityType + Attributable.entityIdAttributeSuffix)) {
                Seq(AttributeString(record.name))
              } else {
                record.toEntity.attributes.get(attributeName) match {
                  case Some(avl: AttributeValueList) =>
                    avl.list.toSeq
                  case Some(av: AttributeValue) =>
                    Seq(av)
                  case _ =>
                    Seq.empty
                }
              }
```
I don't know how frequently method configs specify `${entityType}_id`, but this should be slightly less work in that case

Reply:
Thanks, I was trying to come up with a clever, efficient way to do it and confused myself